# **LAB: TOPIC MODELING**

### **<font color=green>INSTRUCTIONS:</font>** <br>

**<font color=green> 1. Look for EXERCISES in the script (3 in total).</font>** <br>

**<font color=green> 2. Each student INDIVIDUALLY uploads this script with their answers embedded.</font>**

### **Lab Objectives**

1. Learn how to estimate a topic model in Python (using the sklearn package)
2. Get familiar with the output of a topic model
3. Visualize topics in a text corpus
4. Evaluate and discriminate between topic models

### **Session Prep**
Below we install the modules we need and define the text normalization function we used in previous labs, as well as two additional functions we need for today only.

In [2]:
#the module 'sys' allows istalling module from inside Jupyter
import sys

!{sys.executable} -m pip install numpy
import numpy as np

!{sys.executable} -m pip install pandas
import pandas as pd

#Natrual Language ToolKit (NLTK)
!{sys.executable} -m pip install nltk
import nltk

!{sys.executable} -m pip install sklearn
from sklearn import metrics
#from sklearn.model_selection import GridSearchCV
from sklearn.feature_extraction.text import  CountVectorizer #bag-of-words vectorizer
from sklearn.decomposition import LatentDirichletAllocation #package for LDA

# Plotting tools

from pprint import pprint
!{sys.executable} -m pip install pyLDAvis #visualizing LDA
import pyLDAvis
import pyLDAvis.lda_model

import matplotlib.pyplot as plt
%matplotlib inline

#define text normalization function
%run ./Text_Normalization_Function.ipynb #defining text normalization function

#ignore warnings about future changes in functions as they take too much space
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.filterwarnings("ignore", category=DeprecationWarning)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


  Using cached sklearn-0.0.post12.tar.gz (2.6 kB)
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 kB 2.1 MB/s eta 0:00:00
  Created wheel for html.parser: filename=html_parser-0.2-py3-none-any.whl size=1286 sha256=e492d093494b160e5940ad459842dc8c69910d5afc22ae827e2f29588d5039c2
  Stored in directory: /root/.cache/pip/wheels/86/72/1e/d46366a3a2a3843b8b19f0c3c4de3b850247f22b61966de114
Successfully built html.parser


[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.7/23.7 MB 45.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 348.8/348.8 kB 30.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.9/54.9 kB 6.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.3/81.3 kB 11.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100.8/100.8 kB 13.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 137.9/137.9 kB 19.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.6/5.6 MB 85.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.6/101.6 kB 14.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for pattern3: filename=pattern3-3.0.0-py2.py3-none-any.whl size=18554334 sha256=4ca4b25d82e5f62a55693d9ca438c1ff876887f04bc6c29e740b94fe1588f027
  Stored in directory: /root/.cache/pip/wheels/c2/e6/76/468d534

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


Original:   <p>The circus dog in a plissé skirt jumped over Python who wasn't that large, just 3 feet long.</p>
Processed:  ['<', 'p', '>', 'The', 'circus', 'dog', 'in', 'a', 'plissé', 'skirt', 'jumped', 'over', 'Python', 'who', 'was', "n't", 'that', 'large', ',', 'just', '3', 'feet', 'long.', '<', '/p', '>']
Original:   <p>The circus dog in a plissé skirt jumped over Python who wasn't that large, just 3 feet long.</p>
Processed:  <p>The circus dog in a plissé skirt jumped over Python who was not that large, just 3 feet long.</p>


[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


Original:   <p>The circus dog in a plissé skirt jumped over Python who wasn't that large, just 3 feet long.</p>
Processed:  [('<', 'a'), ('p', 'n'), ('>', 'v'), ('the', None), ('circus', 'n'), ('dog', 'n'), ('in', None), ('a', None), ('plissé', 'n'), ('skirt', 'n'), ('jumped', 'v'), ('over', None), ('python', 'n'), ('who', None), ('was', 'v'), ("n't", 'r'), ('that', None), ('large', 'a'), (',', None), ('just', 'r'), ('3', None), ('feet', 'n'), ('long.', 'a'), ('<', 'n'), ('/p', 'n'), ('>', 'n')]
Original:   <p>The circus dog in a plissé skirt jumped over Python who wasn't that large, just 3 feet long.</p>
Processed:  < p > the circus dog in a plissé skirt jump over python who be n't that large , just 3 foot long. < /p >
Original:   <p>The circus dog in a plissé skirt jumped over Python who wasn't that large, just 3 feet long.</p>
Processed:    p   The circus dog in a plissé skirt jumped over Python who was n t that large   just 3 feet long     p  
Original:   <p>The circus dog in a pli

Below we define two functions that will display the results of fitting a topic model, to be used later:

**Note**: these functions are not the focus of the lab, therefore we'll not be discussing them, but you are welcome to explore and dig into them later if you prefer.

In [3]:
def display_topics(model, feature_names, no_top_words):
    for topic_idx, topic in enumerate(model.components_):
        print("Topic %d:" % (topic_idx))
        print(" ".join([feature_names[i]
                        for i in topic.argsort()[:-no_top_words - 1:-1]]))

def get_topic_words(vectorizer, lda_model, n_words):
    keywords = np.array(vectorizer.get_feature_names_out())
    topic_words = []
    for topic_weights in lda_model.components_:
        top_word_locs = (-topic_weights).argsort()[:n_words]
        topic_words.append(keywords.take(top_word_locs).tolist())
    return topic_words

### **Toy Data Example**
We'll start with working on a toy dataset. It will allow us to grasp the full results of a topic model before moving on to high-dimensional realistic data.

#### **Define and Prep Toy Data**
Let's use the toy corpus on animals and programming similar to one in the lecture. Let's define it:

In [4]:
toy_corpus = ["The fox jumps over the dog",
              "The fox is very clever and quick",
              "The dog is slow and lazy",
              "The cat is smarter than the fox and the dog but it can never learn Java",
              "Python is an excellent programming language",
              "Java and Ruby are other programming languages",
              "Python and Java are very popular programming languages",
              "Python programs are smaller than Java programs"]

Let's **normalize** our toy_corpus and call the normalized corpus **normalized_toy_corpus**:

In [5]:
normalized_toy_corpus = normalize_corpus(toy_corpus)

Since for topic modeling we need text data in the **Bag-of-Words** representation, let's **vectorize** our normalized_toy_corpus and call it **bow_toy_corpus**:

In [6]:
#define the bag-of-words vectorizer:
bow_vectorizer = CountVectorizer()

#vectorize the normalized data:
bow_toy_corpus = bow_vectorizer.fit_transform(normalized_toy_corpus)

Have a look at the Bag-of-Words representation of our corpus: **It never hurts to know how you data look like :)** Note absence of stopwords and other differences with the raw data:

In [7]:
pd.DataFrame(data = bow_toy_corpus.todense(), columns = bow_vectorizer.get_feature_names_out())

cat  clever  dog  excellent  fox  java  jump  language  lazy  learn  never  \
0    0       0    1          0    1     0     1         0     0      0      0   
1    0       1    0          0    1     0     0         0     0      0      0   
2    0       0    1          0    0     0     0         0     1      0      0   
3    1       0    1          0    1     1     0         0     0      1      1   
4    0       0    0          1    0     0     0         1     0      0      0   
5    0       0    0          0    0     1     0         1     0      0      0   
6    0       0    0          0    0     1     0         1     0      0      0   
7    0       0    0          0    0     1     0         0     0      0      0   

   popular  program  programming  python  quick  ruby  slow  small  smarter  
0        0        0            0       0      0     0     0      0        0  
1        0        0            0       0      1     0     0      0        0  
2        0        0            0       0      0     0     1      0        0  
3        0        0            0       0      0     0     0      0        1  
4        0        0            1       1      0     0     0      0        0  
5        0        0            1       0      0     1     0      0        0  
6        1        0            1       1      0     0     0      0        0  
7        0        2            0       1      0     0     0      1        0

#### **Topic Model (via Latent Dirichlet Allocation) on Toy Data**
Now let's **model topics** in our toy data. Given that the toy corpus is so small, we know all "topics" it contains (**what are they?**) and it will be easy for us:<br> 1) to check if the topic model results make sense; <br>2) see all the results that the topic model produces.  <br><br>
We will be using the **LatentDirichletAllocation** function which we already imported earlier (see Session Prep). The function has the following **parameters** to be set:
1. Number of topics to model: **n_components**
2. Parameter vector for the Dirichlet distribution for *topics*: **doc_topic_prior**
3. Parameter vector for the Dirichlet distribution for *words* in a topic: **topic_word_prior**

Notes on **parameter vectors for the Dirichlet distributions**: <br>
1. Although the Dirichlet distribution parameters are represented by a **vector**, for simplicity we provide one number for each parameter vector. For example, if we set the number of topics to 2 (n_components=2), the parameter vector for the Dirichlet distribution for *topics* should be a two-dimensional vector. We set doc_topic_prior=0.5 and the LatentDirichletAllocation function internally creates a two-dimensional vector (0.5,0.5). Similar logic applies to the parameter vector for the Dirichlet distribution for *words* (the dimensionality of that parameter vector is equal to the number of terms in the text corpus, which is typically very large).<br><br>
2. Remember, that we need **sparsity** in the distribution of topics across documents (i.e., some documents have a zero probability of containing some of the topics) and *sparsity* in the distribution of words in topics (i.e., some words have zero probability to be present in some topics). To induce sparsity, we need to set **doc_topic_prior** and **topic_word_prior** between 0 and 1.

Now, let's set the parameters and estimate the topic model:

In [8]:
lda_toy_corpus = LatentDirichletAllocation(n_components=2, max_iter=500,
                                           doc_topic_prior = 0.9,
                                           topic_word_prior = 0.9).fit(bow_toy_corpus)

Display results by showing 15 **most frequent (top)** words for each topic (we use **function display_topics** defined in Session Prep):

In [9]:
no_top_words = 15
display_topics(lda_toy_corpus, bow_vectorizer.get_feature_names_out(), no_top_words)

Topic 0:
dog fox smarter never learn cat jump lazy slow quick clever java programming language python
Topic 1:
java python language programming program popular small excellent ruby fox dog clever quick lazy slow


Note that topics do not have names or labels. **Topics are just collections of words**, following the definition of a topic in text mining. <br><br> To be precise, topics are **word vectors**, where each vector element is the **weight** (relative frequency) of the word in a topic. Let's have a look at those "word vectors". Can you see below that each word vector (topic) is a **simplex**?

In [10]:
word_weights = lda_toy_corpus.components_ / lda_toy_corpus.components_.sum(axis=1)[:, np.newaxis]
pd.DataFrame(word_weights.T, index = bow_vectorizer.get_feature_names_out()).T

cat    clever       dog  excellent       fox      java      jump  \
0  0.055088  0.054435  0.113486   0.028202  0.113486  0.053816  0.054604   
1  0.025956  0.026564  0.026766   0.050981  0.026766  0.109888  0.026407   

   language      lazy     learn     never   popular   program  programming  \
0  0.028622  0.054435  0.055088  0.055088  0.027959  0.028292     0.028622   
1  0.105755  0.026564  0.025956  0.025956  0.051208  0.078480     0.105755   

     python     quick      ruby      slow     small   smarter  
0  0.028476  0.054435  0.028333  0.054435  0.028007  0.055088  
1  0.105892  0.026564  0.050859  0.026564  0.051163  0.025956

### **<font color=green>EXERCISE 1:</font>**

**<font color=green>1.1. Adjust topic_word_prior (beta) and observe the changes in topic representation. Try these values: 1) topic_word_prior = 0.1; 2) topic_word_prior = 1. What changes in the topic model results do you observe in these two cases? Why? </font>**<br>

<font color=green> **Note**: You can also adjust doc_topic_prior, but you will be able to see the results of this change when we compute shares of topics by document later </font> <br><br>

**<font color=green>1.2. (Question) You are likely to be not 100% satisfied with the model performance. The word "java" might still be appearing among the top words in the "animals" topic. Why? Looking at the text corpus might help to find the answer.</font>**

**Answer 1.1:** <br>


 topic_word_prior = 0.1:

In [11]:
lda_toy_corpus = LatentDirichletAllocation(n_components=2, max_iter=500,
                                           doc_topic_prior = 0.9,
                                           topic_word_prior = 0.1).fit(bow_toy_corpus)
no_top_words = 15
display_topics(lda_toy_corpus, bow_vectorizer.get_feature_names_out(), no_top_words)
word_weights = lda_toy_corpus.components_ / lda_toy_corpus.components_.sum(axis=1)[:, np.newaxis]
pd.DataFrame(word_weights.T, index = bow_vectorizer.get_feature_names_out()).T

Topic 0:
dog fox jump lazy slow popular clever quick smarter learn never cat java language programming
Topic 1:
java python programming language program small ruby excellent smarter learn never cat quick clever popular


cat    clever       dog  excellent       fox      java      jump  \
0  0.008337  0.008338  0.258326   0.008334  0.258325  0.008335  0.091664   
1  0.042306  0.042306  0.003847   0.042308  0.003847  0.157693  0.003846   

   language      lazy     learn     never   popular   program  programming  \
0  0.008334  0.091664  0.008337  0.008337  0.091659  0.008334     0.008334   
1  0.119232  0.003846  0.042306  0.042306  0.003849  0.080770     0.119232   

     python     quick      ruby      slow     small   smarter  
0  0.008334  0.008338  0.008334  0.091664  0.008334  0.008337  
1  0.119232  0.042306  0.042308  0.003846  0.042308  0.042306

 topic_word_prior = 1:

In [12]:
lda_toy_corpus = LatentDirichletAllocation(n_components=2, max_iter=500,
                                           doc_topic_prior = 0.9,
                                           topic_word_prior = 1).fit(bow_toy_corpus)
no_top_words = 15
display_topics(lda_toy_corpus, bow_vectorizer.get_feature_names_out(), no_top_words)
word_weights = lda_toy_corpus.components_ / lda_toy_corpus.components_.sum(axis=1)[:, np.newaxis]
pd.DataFrame(word_weights.T, index = bow_vectorizer.get_feature_names_out()).T

Topic 0:
dog fox smarter never learn cat java jump lazy slow quick clever programming language python
Topic 1:
java python language programming program popular small excellent ruby fox dog clever quick lazy slow


cat    clever       dog  excellent       fox      java      jump  \
0  0.054639  0.053903  0.109535   0.029647  0.109535  0.054114  0.054112   
1  0.027375  0.028062  0.028375   0.050714  0.028375  0.106265  0.027866   

   language      lazy     learn     never   popular   program  programming  \
0  0.030207  0.053903  0.054639  0.054639  0.029384  0.029815     0.030207   
1  0.102458  0.028062  0.027375  0.027375  0.050959  0.076691     0.102458   

     python     quick      ruby      slow     small   smarter  
0  0.030042  0.053903  0.029793  0.053903  0.029445  0.054639  
1  0.102612  0.028062  0.050578  0.028062  0.050902  0.027375

  When the "topic_word_prior" increases, the sparsity decreases, so the weight of each word become more similar.

**Answer 1.2:**

 In the line "The cat is smarter than the fox and the dog but it can never learn Java," there are three animal-related terms: cat, fox, and dog. Thus, machine might categorize Java as a similar term.

### **Topic Modeling on Real Data**

The dataset here is the one we used for doing Text Classification in Lab 3. The newspaper blog posts have 4 topics: **atheism, religion, computer graphics, and space science**. Of course, we will *not use* class label information for topic modeling.

Download the data and set up the data (**news_corpus**):

In [13]:
from sklearn.datasets import fetch_20newsgroups
categories = ['alt.atheism', 'talk.religion.misc', 'comp.graphics', 'sci.space']
dataset = fetch_20newsgroups(shuffle=True,
                             random_state=1,
                             categories = categories,
                             remove=('headers', 'footers', 'quotes'))
news_corpus = dataset.data

Now, let's normalize the corpus and create the Bag-of-Words representation of the data. We'll limit the number of features to **1000 most frequent features** to compute the topic model faster.

In [14]:
#normalize data
normalized_corpus_news = normalize_corpus(news_corpus)

#define a Bag-of-Words vecgtorizer
bow_vectorizer_news = CountVectorizer(max_features=1000)

#vectorize data
bow_news_corpus = bow_vectorizer_news.fit_transform(normalized_corpus_news)

Now let's fit the topic model. We need to **set the number of topics** first. We are *lucky to know* that there are **4 topics** (atheism, religion, computer graphics, and space science) and it will allow us to judge the performance of the topic model better.

**Note**: It will take a couple of minutes for the estimation to finish. The larger the number of iterations (max_iter) you allow for, the longer it takes. You can also find that you might get a slightly different result (for the estimated weights) if you rerun the model (qualitatively, the result is likely to stay the same). This behavior is due to the stochastic nature of the model. Estimation starts from a random guess for the weights and iteratively improves. When we do not allow for a large number of iterations, we are more likely to be affected by the random starting values.

In [15]:
lda_news = LatentDirichletAllocation(n_components=4, max_iter=100,
                                     doc_topic_prior = 0.25,
                                     topic_word_prior = 0.25).fit(bow_news_corpus)

Display results with top 10 words for each topic:

In [16]:
no_top_words_news = 10
display_topics(lda_news, bow_vectorizer_news.get_feature_names_out(), no_top_words_news)

Topic 0:
god jesus christian believe religion bible atheist argument belief atheism
Topic 1:
image file use edu program software graphic format jpeg ftp
Topic 2:
think people like know could good well time thing point
Topic 3:
space nasa launch satellite orbit mission earth year system use


Display **word vectors** (words are in alphabetical order) for each topic. Each column is a topic:

In [17]:
word_weights = lda_news.components_ / lda_news.components_.sum(axis=1)[:, np.newaxis]
word_weights_df = pd.DataFrame(word_weights.T,
                               index = bow_vectorizer_news.get_feature_names_out(),
                               columns = ["Topic_" + str(i) for i in range(4)])
word_weights_df.head(10)

Topic_0   Topic_1   Topic_2   Topic_3
3d          0.000010  0.005132  0.000006  0.000010
able        0.000011  0.001311  0.002007  0.000265
ac          0.000013  0.001246  0.000040  0.000398
accept      0.002539  0.000314  0.001576  0.000011
acceptable  0.000012  0.000010  0.000841  0.000319
access      0.000011  0.001905  0.000246  0.000932
accord      0.001565  0.000009  0.000898  0.000248
account     0.001005  0.000055  0.000426  0.000102
act         0.003300  0.000010  0.000424  0.000933
action      0.000395  0.000037  0.002103  0.000129

Now, **sort by word weights in Topic 0** (descending order) and see the weights by 10 most frequent words in Topic 0:

In [18]:
word_weights_df.sort_values(by='Topic_0',ascending=False).head(10)

Topic_0   Topic_1   Topic_2   Topic_3
god        0.031887  0.000045  0.000007  0.000010
jesus      0.016595  0.000009  0.000006  0.000010
christian  0.014860  0.000009  0.000006  0.000010
believe    0.012084  0.000010  0.002576  0.000585
religion   0.012074  0.000009  0.000108  0.000010
bible      0.011591  0.000009  0.000006  0.000010
atheist    0.010380  0.000009  0.000006  0.000010
argument   0.008928  0.000009  0.000707  0.000010
belief     0.008403  0.000009  0.000007  0.000010
atheism    0.008080  0.000009  0.000006  0.000010

### **Topic Model Visualization**

You can **visualize** the topics: topic size, frequency of words in a topic and so on.

In this visualization, you can rank words in a topic by **relevancy**: do you want rare and exclusive terms (i.e. found mostly in that topic) or terms that are used frequently in that topic, but not always exclusive to that topic? Relevancy parameter is λ (0 ≤ λ ≤ 1). You can adjust it:

* small λ highlights potentially rare, but exclusive terms for the selected topic;
* large values of λ (near 1) highlight frequent, but not necessarily exclusive, terms for the selected topic;

Formally, relevancy is measured using the following formula:

    Relevancy = λ log[p(term | topic)] + (1 - λ) log[p(term | topic)/p(term)],
   
   where p(term | topic) stands for word weight in a topic and p(term) stands for word's weight in a corpus.

Additional information on how to use this visualization:
* http://www.kennyshirley.com/LDAvis/
* https://nlp.stanford.edu/events/illvi2014/papers/sievert-illvi2014.pdf

We installed all the **pyLDAvis** module required for this visualization in Session Prep. Now let's use it:

In [19]:
#prepare to display result in the Jupyter notebook
pyLDAvis.enable_notebook()

#run the visualization [mds is a function to use for visualizing the "distance" between topics]
pyLDAvis.lda_model.prepare(lda_news, bow_news_corpus, bow_vectorizer_news, mds='tsne')

PreparedData(topic_coordinates=               x           y  topics  cluster       Freq
topic                                                   
2      18.091793  -18.516022       1        1  33.700703
1      45.352833 -101.601265       2        1  24.041231
0     -10.112836  -73.648613       3        1  21.151535
3      73.544106  -46.387661       4        1  21.106531, topic_info=        Term         Freq        Total Category  logprob  loglift
846    space  1043.000000  1043.000000  Default  30.0000  30.0000
368      god   788.000000   788.000000  Default  29.0000  29.0000
416    image   845.000000   845.000000  Default  28.0000  28.0000
329     file   546.000000   546.000000  Default  27.0000  27.0000
576     nasa   413.000000   413.000000  Default  26.0000  26.0000
..       ...          ...          ...      ...      ...      ...
943      use   206.689716  1087.709886   Topic4  -4.7803  -0.1050
589      new   152.997438   377.598587   Topic4  -5.0811   0.6522
787  science   129.448056   237.106585   Topic4  -5.2483   0.9504
421  include   134.283890   381.556436   Topic4  -5.2116   0.5113
913     time   135.388555   652.472516   Topic4  -5.2034  -0.0170

[277 rows x 6 columns], token_table=      Topic      Freq      Term
term                           
0         2  0.995058        3d
11        1  0.887521  actually
11        2  0.068677  actually
11        3  0.042263  actually
25        3  0.982859     allah
...     ...       ...       ...
996       1  0.467771      year
996       3  0.013624      year
996       4  0.519998      year
997       1  0.964492       yes
997       3  0.028791       yes

[469 rows x 3 columns], R=30, lambda_step=0.01, plot_opts={'xlab': 'PC1', 'ylab': 'PC2'}, topic_order=[3, 2, 1, 4])

### **<font color=green>EXERCISE 2:</font>**
    
**<font color=green>2.1. Fit a topic model with 3 topics (n_components = 3). The script is provided below. Important! Note that the model with three topics is called lda_news_3_topics.</font>**

**<font color=green>2.2. Use the visualization tool to answer the following question: Which topic is the most common / largest topic in the corpus? Can you give a name to that topic? List 5 most relevant and exclusive terms for that topic (with lambda = 0.2).</font>**

**<font color=green>2.3. You fit the model with 3 topics, but you know that the dataset has four classes (topics). Which 2 of the 4 classes ('atheism', 'religion', 'computer graphics', 'space science') were grouped together into one topic by the topic model when you fit it with 3 topics? Why?</font>**

Your answer (need to add lines of code related to visualization):

**Answer 2.1**:

Code:

In [21]:
#fit the LDA model with 3 topics
lda_news_3_topics = LatentDirichletAllocation(n_components=3, max_iter=100,
                                     doc_topic_prior = 0.25,
                                     topic_word_prior = 0.25).fit(bow_news_corpus)


**Answer 2.2:**

In [23]:
no_top_words_news = 10

word_weights = lda_news_3_topics.components_ / lda_news_3_topics.components_.sum(axis=1)[:, np.newaxis]
word_weights_df = pd.DataFrame(word_weights.T,
                               index = bow_vectorizer_news.get_feature_names_out(),
                               columns = ["Topic_" + str(i) for i in range(3)])

word_weights_df.sort_values(by='Topic_0',ascending=False).head(10)

#prepare to display result in the Jupyter notebook
pyLDAvis.enable_notebook()

#run the visualization [mds is a function to use for visualizing the "distance" between topics]
pyLDAvis.lda_model.prepare(lda_news_3_topics, bow_news_corpus, bow_vectorizer_news, mds='tsne')

PreparedData(topic_coordinates=                 x            y  topics  cluster       Freq
topic                                                      
0      1290.562622  1098.234253       1        1  37.749610
2      -456.176208 -1325.102905       2        1  35.701556
1     -3191.051025  -255.021286       3        1  26.548834, topic_info=        Term         Freq        Total Category  logprob  loglift
846    space  1042.000000  1042.000000  Default  30.0000  30.0000
416    image   850.000000   850.000000  Default  29.0000  29.0000
368      god   787.000000   787.000000  Default  28.0000  28.0000
209     data   444.000000   444.000000  Default  27.0000  27.0000
479   launch   398.000000   398.000000  Default  26.0000  26.0000
..       ...          ...          ...      ...      ...      ...
972      way   174.346223   434.154690   Topic3  -5.1799   0.4138
290     even   160.300622   457.092959   Topic3  -5.2639   0.2783
700  program   166.091350   532.675460   Topic3  -5.2284   0.1608
648   people   174.296794   790.158892   Topic3  -5.1802  -0.1853
679     post   148.759525   397.820015   Topic3  -5.3386   0.3425

[207 rows x 6 columns], token_table=      Topic      Freq     Term
term                          
17        2  0.082318  advance
17        3  0.919222  advance
48        1  0.029887  anybody
48        3  0.971314  anybody
49        1  0.225618   anyone
...     ...       ...      ...
989       2  0.287673     work
989       3  0.553771     work
996       1  0.104561     year
996       2  0.322776     year
996       3  0.570541     year

[308 rows x 3 columns], R=30, lambda_step=0.01, plot_opts={'xlab': 'PC1', 'ylab': 'PC2'}, topic_order=[1, 3, 2])

**Answer 2.3:**

By my observation, computer graphics and space science were grouped together into one topic by the topic model when you fit it with 3 topics.
The reason is that both topics have a high similarity in their texts as both are in the science field.

### **How To Find Dominant Topic in a Document**

Each document contains several topics. One of the topics is **dominant**, i.e. it is the largest topic in the document. That topic gives you an answer to the question: **What is this document about?** In other words, the document's dominant topic **summarizes** the document.

Let's assign a dominant topic to **each document** in our corpus. Weights in a word vector for a topic provide a measure of association for the word with the topic. If you sum weights for a particular topic across all words in a document, you'll get the weight of that topic in the document.

The attribute **.transform** to our function **lda_news** computes the weights of each topic in documents:

In [24]:
lda_news_topic_weights = lda_news.transform(bow_news_corpus)

Let's convert lda_news_topic_weights into a nice-looking dataframe and have a look at the computed topic weights in documents:

In [25]:
#array of document "names" and topic "names" ("names" are just indecies)
doc_names = ["Doc_" + str(i) for i in range(len(normalized_corpus_news))]
topic_names = ["Topic_" + str(i) for i in range(4)]

#convert to dataframe
df_document_topic = pd.DataFrame(np.round(lda_news_topic_weights, 4), columns=topic_names, index=doc_names)
df_document_topic.head(5)

Topic_0  Topic_1  Topic_2  Topic_3
Doc_0   0.9986   0.0004   0.0005   0.0005
Doc_1   0.0026   0.5920   0.4025   0.0029
Doc_2   0.6490   0.0106   0.3297   0.0106
Doc_3   0.0108   0.7622   0.0107   0.2162
Doc_4   0.0417   0.8738   0.0424   0.0422

The topic with the highest weight in each document is a **dominant topic**. The weights across the 4 topics sum up to 1. Let's add a column that shows dominant topic for each document:

In [26]:
#vector of indecies for columns with the highest value by each row in df_document_topic
dominant_topic = np.argmax(df_document_topic.values, axis=1)

#add dominant_topic as a column to df_document_topic
df_document_topic['dominant_topic'] = dominant_topic
df_document_topic.head(5)

Topic_0  Topic_1  Topic_2  Topic_3  dominant_topic
Doc_0   0.9986   0.0004   0.0005   0.0005               0
Doc_1   0.0026   0.5920   0.4025   0.0029               1
Doc_2   0.6490   0.0106   0.3297   0.0106               0
Doc_3   0.0108   0.7622   0.0107   0.2162               1
Doc_4   0.0417   0.8738   0.0424   0.0422               1

### **Topic Model Evaluation: Log-likelihood, Perplexity and Coherence Scores**

Log-likelihood, Perplexity and Coherence Score are **measures of performance** for a topic model. They are used for comparing and discriminating between topic models estimated on the same data. Log-likelihood, perplexity and coherence scores **do not have** a baseline or a threshold values and therefore are useful only for comparing models.

How do you specify different models? You can set **different number of topics** and also play with the **parameters of the Dirichlet distributions**.

#### **Coherence Score**

We will use a function **CoherenceModel()** from the **gensim** module (you can also explore that package as it can be used to estimate an LDA model). The sklearn module does not have the functionality to compute the coherence score. Let's install the gensim package and the functions needed:

In [27]:
!{sys.executable} -m pip install gensim
import gensim

from gensim.models.coherencemodel import CoherenceModel
from gensim.corpora.dictionary import Dictionary

The function CoherenceModel() needs as **inputs**:

**1. Dictionary of the corpus**<br>
**2. Corpus with each document represented as Bag-of-Words**<br>
**3. An array of top words for each topic: we'll have top 20 words for each topic**
  
We will now create those objects:

In [28]:
#tokenizing the corpus
news_corpus_tokenized = [tokenize_text(normalized_corpus_news[doc_id]) for doc_id in range(len(normalized_corpus_news))]

#Dictionary of the corpus:
news_dictionary = Dictionary(news_corpus_tokenized)

#Bag-of-words representation for each document of the corpus:
news_corpus_bow = [news_dictionary.doc2bow(doc) for doc in news_corpus_tokenized]

#top 20 words for each topic (using the function defined in session prep)
topic_topwords = get_topic_words(vectorizer = bow_vectorizer_news, lda_model = lda_news, n_words=20)

Now let's compute **the coherence score for the model overall**. We use one of the coherence metrics "u-mass" which measures semantic similarity of words in a topic, but there are other metrics as well.

**Note**: You can check out different coherence metrics here if you are interested: https://dl.acm.org/doi/abs/10.1145/2684822.2685324

In [29]:
cm = CoherenceModel(topics=topic_topwords,
                    corpus = news_corpus_bow ,
                    dictionary = news_dictionary, coherence='u_mass')
print("Coherence score for the model: ", np.round(cm.get_coherence(), 4))  # get coherence value

Coherence score for the model:  -1.4478


You can also see **coherence scores by topic**:

In [30]:
print("Coherence score by topic (higher values are better): ", np.round(cm.get_coherence_per_topic(),4))

Coherence score by topic (higher values are better):  [-1.379  -1.7375 -1.32   -1.3548]


#### **Log-Likelihood Score**

To compute the log-likelihood score we use the **.score** attribute of our defined and fitted LDA function:

In [31]:
print("Log-Likelihood (higher values are better): ", lda_news.score(bow_news_corpus))

Log-Likelihood (higher values are better):  -741213.9353004873


#### **Perplexity Score**

To compute the Perplexity score we use the **.perplexity** attribute of our defined and fitted LDA function:

In [32]:
print("Perplexity (lower values are better): ", lda_news.perplexity(bow_news_corpus))

Perplexity (lower values are better):  574.0715741855637


### **<font color=green>EXERCISE 3</font>**

**<font color=green>Compare the coherence score, perplexity score and the log-likelihood for models with 2, 3, and 4 topics with your human-judgment-based evaluation of those models. What do you find? </font>**

**<font color=green>What you need to do:</font>**

**<font color=green>3.1. (Question) For model with 4 topics - All code work is done: The model and evaluation metrics are already computed above. You just need to look up the values for the coherence, perplexity and log-likelihood for the model with 4 topics above and discuss what you observe. You might be interested in looking at coherence score by topic as well;</font>**

**<font color=green>3.2. For model with 3 topics - The model is computed in Exercise 2. You need to compute the perplexity, log-likelihood and coherence scores for the model with 3 topics (the lines for the coherence score are provided below) and dicuss your results;</font>**

**<font color=green>3.3. For model with 2 topics - You need to fit the model with 2 topics and compute all 3 evaluation metrics; dicuss your results.</font>**

**Answer 3.1:**

Coherence score for the model:  -1.4478

Coherence score by topic (higher values are better):  [-1.379  -1.7375 -1.32   -1.3548]

Log-Likelihood (higher values are better):  -741213.9353004873

Perplexity (lower values are better):  574.0715741855637

Since there are now comparisons, it's hard to interpret the above values.

Thus, we should observe the computed topic weights in documents.

For model with 4 topics, every document's dominant topic has a probability larger than 50%. Which indicate that the 4-topic-model might be accurate enough.


**Answer 3.2:**

<font color=blue>Code (complete the lines):<font>

In [38]:
#coherence score for 3 topics:
topic_topwords_3_topics = get_topic_words(vectorizer = bow_vectorizer_news, lda_model = lda_news_3_topics, n_words=20)
cm_3_topics = CoherenceModel(topics=topic_topwords_3_topics,
                             corpus = news_corpus_bow,
                             dictionary = news_dictionary, coherence='u_mass')
#overall coherence score for the model:
print("Coherence score for the model: (higher values are better)", np.round(cm_3_topics.get_coherence(), 3))
print("Coherence score by topic (higher values are better): ", np.round(cm.get_coherence_per_topic(),3))
print("Log-Likelihood (higher values are better): ", lda_news_3_topics.score(bow_news_corpus))
print("Perplexity (lower values are better): ", lda_news_3_topics.perplexity(bow_news_corpus))

Coherence score for the model: (higher values are better) -1.515
Coherence score by topic (higher values are better):  [-1.379 -1.737 -1.32  -1.355]
Log-Likelihood (higher values are better):  -748864.0845707384
Perplexity (lower values are better):  612.973401861286


Coherence score and log-likelihood become lower, and perplexity becomes higher. Which shows the 3-topic-model is worse than 4-topic-model.


**Answer 3.3:**

Code:

In [37]:
#fit the LDA model with 2 topics
lda_news_2_topics = LatentDirichletAllocation(n_components=2, max_iter=100,
                                     doc_topic_prior = 0.25,
                                     topic_word_prior = 0.25).fit(bow_news_corpus)

#coherence score for 2 topics:
topic_topwords_2_topics = get_topic_words(vectorizer = bow_vectorizer_news, lda_model = lda_news_2_topics, n_words=20)
cm_2_topics = CoherenceModel(topics=topic_topwords_2_topics,
                             corpus = news_corpus_bow,
                             dictionary = news_dictionary, coherence='u_mass')

#overall coherence score for the model:
print("Coherence score for the model: (higher values are better)", np.round(cm_2_topics.get_coherence(), 2))
print("Coherence score by topic (higher values are better): ", np.round(cm.get_coherence_per_topic(),2))
print("Log-Likelihood (higher values are better): ", lda_news_2_topics.score(bow_news_corpus))
print("Perplexity (lower values are better): ", lda_news_2_topics.perplexity(bow_news_corpus))

Coherence score for the model: (higher values are better) -1.42
Coherence score by topic (higher values are better):  [-1.38 -1.74 -1.32 -1.35]
Log-Likelihood (higher values are better):  -751871.2002129115
Perplexity (lower values are better):  628.9770439681971


Coherence score and log-likelihood become lower, and perplexity becomes higher. Which shows the 2-topic-model is worse than 3-topic-model.

<br>**NOTE:** Generally, you can write a simple script that selects the best topic model **automatically** based on a criterion for "best model" (log-likelihood, perplexity, or coherence score). The script can vary both parameters of the Dirichlet distributions and the number of topics, or just the number of topics.